yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -256,6 +256,12 @@ private[spark] class ApplicationMaster(
 
   private def runDriver(securityMgr: SecurityManager): Unit = {
     addAmIpFilter()
+
+    // This must be done before SparkContext is initialized, since the CoarseGrainedSchedulerBackend
+    // is started at that time. That is what schedules the re-logins. It is scheduled only if the
+    // principal is actually setup. So we make sure it is available.
+    SparkHadoopUtil.get.setPrincipalAndKeytabForLogin(
+      System.getenv("SPARK_PRINCIPAL"), System.getenv("SPARK_KEYTAB"))
```

Comment:
I am actually following the same method Yarn module already uses. The spark staging directory is passed in as an environment variable. I don't have a preference for either. I will switch it out to spark.yarn.principal and spark.yarn.keytab